src/Bicep.Types/Concrete/ResourceType.cs

```diff
@@ -11,15 +11,51 @@ namespace Azure.Bicep.Types.Concrete
     public enum ResourceFlags
     {
         None = 0,
-        ReadOnly = 1 << 0,
-        WriteOnly = 1 << 1,
+        ReadOnly = 1 << 0
     }
 
     public class ResourceType : TypeBase
     {
         [JsonConstructor]
-        public ResourceType(string name, ScopeType scopeType, ScopeType? readOnlyScopes, ITypeReference body, ResourceFlags flags, IReadOnlyDictionary<string, ResourceTypeFunction>? functions)
-            => (Name, ScopeType, ReadOnlyScopes, Body, Flags, Functions) = (name, scopeType, readOnlyScopes, body, flags, functions);
+        public ResourceType(
+            string name, 
+            ScopeType scopeType, 
+            ScopeType? readOnlyScopes, 
+            ITypeReference body, 
+            ResourceFlags flags, 
+            IReadOnlyDictionary<string, ResourceTypeFunction>? functions, 
+            ScopeType? writableScopes = null, 
+            ScopeType? readableScopes = null)
+        {
+            // Check for illegal mixing of legacy and new scope fields
+            bool hasLegacy = scopeType != ScopeType.Unknown ||
+                 (readOnlyScopes.HasValue && readOnlyScopes.Value != ScopeType.Unknown) ||
+                 flags != ResourceFlags.None;
+
+            bool hasModern =
+                (writableScopes.HasValue && writableScopes.Value != ScopeType.Unknown) ||
```

Comment:
`ScopeType.Unknown` isn't meant to be a sentinel of whether the field was set or not. I would just rely on `.HasValue` here